lib/Doctrine/Annotations/Metadata/AnnotationMetadata.php

```diff
@@ -0,0 +1,97 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Doctrine\Annotations\Metadata;
+
+use function array_combine;
+use function array_filter;
+use function array_map;
+use function array_values;
+use function assert;
+use function count;
+
+/**
+ * @internal
+ */
+final class AnnotationMetadata
+{
+    /** @var string */
+    private $name;
+
+    /** @var AnnotationTarget */
+    private $target;
+
+    /** @var bool */
+    private $usesConstructor;
+
+    /** @var PropertyMetadata[] */
+    private $properties;
+
+    /** @var PropertyMetadata|null */
+    private $defaultProperty;
+
+    /**
+     * @param PropertyMetadata[] $properties
+     */
+    public function __construct(
+        string $name,
+        AnnotationTarget $target,
+        bool $hasConstructor,
+        PropertyMetadata ...$properties
+    ) {
+        $this->name           = $name;
+        $this->target         = $target;
+        $this->usesConstructor = $hasConstructor;
+        $this->properties     = array_combine(
+            array_map(
+                static function (PropertyMetadata $property) : string {
+                    return $property->getName();
+                },
+                $properties
+            ),
+            $properties
+        );
+
+        $defaultProperties = array_values(
+            array_filter(
+                $properties,
+                static function (PropertyMetadata $property) : bool {
+                    return $property->isDefault();
+                }
+            )
+        );
+
+        assert(count($defaultProperties) <= 1);
```

Comment:
I wouldn't use `assert` here. To quote from the manual:

> Assertions should not be used for normal runtime operations like input parameter checks. As a rule of thumb your code should always be able to work correctly if assertion checking is not activated.

I agree with that statement and would rather throw a subclass of `InvalidArgumentException` instead of using `assert`: while the code does not fail now if there is more than one default property, this can introduce issues that are difficult to debug at a later stage.